Migration: add submission date to requests table

zakieh-tayyebi opened the pull request:
## Migrating address fields
Address fields will no longer be migrated to the households table, but to the "Requests" and "Social Service Requests" tables, only for furniture and MESH requests. Furniture requests will include only `Geocode`. MESH requests will include `Street Address`, `City, State`, `Zip Code`, and `Geocode.`

## Migrating the request submission date
Currently, the "Households" table has `Legacy First Date Submitted` and  `Legacy Last Date Submitted` for all open requests for each household. However, we need `Legacy Date Submitted` in the "Requests" and "Social Service Requests" tables to know the date per open request.

- Modified function `transform_open_requests()` to keep open request items as a `pandas.DataFrame` object containing the `Date Submitted` column for each open request.
- For each type of item, the oldest open request is selected, so the column `item` is unique in each table.
- The records migrated to the "Requests" and "Social Service Requests" tables in the new base have `Legacy Date Submitted` set to `Date Submitted` in the old base.
- In the "Requests" and "Social Service Requests" tables, Airtable automation sets `Request Opened At` to `Legacy Date Submitted` if it's not empty (for migrated requests), and otherwise to `CREATED_TIME()` (for requests submitted to the new base).

zakieh-tayyebi committed:
added Legacy Date Submitted to requests table

zakieh-tayyebi committed:
making output a dataframe type

zakieh-tayyebi committed:
fixed dataframe issues

zakieh-tayyebi committed:
removing duplicates and picking oldest request of each type

zakieh-tayyebi commented on core/scripts/migrate_requests_to_new_base.py:
```diff
@@ -382,29 +390,30 @@ def transform_open_requests(
-    )
+    def item_df(r, item):
+        df = pd.DataFrame({"item": [item]})
+        df.insert(column=DATE_SUBMITTED_FIELD, value=date_parser.parse(r.get(DATE_SUBMITTED_FIELD, [])), loc = 0)
```
@abelsonlive Is there a better function to parse the "Date Submitted" column date format? Right now using `dateutil.date_parser.parse` without specifying a specific format.

zakieh-tayyebi replied:
Actually I removed the parser since `transform_date_submitted` sorts dates with their untransformed format.

zakieh-tayyebi committed:
removed parser

sunwsusan commented on core/scripts/migrate_requests_to_new_base.py:
```diff
@@ -1,6 +1,9 @@
 import copy
+from datetime import datetime
+from dateutil import parser as date_parser
+import pandas as pd
```
`pandas` is not a native python library.  please add to `requirements.txt` file here: https://github.com/bushwickayudamutua/bam-automation/blob/main/core/scripts/requirements.txt

zakieh-tayyebi replied:
great catch! Added it.

zakieh-tayyebi committed:
removed commented part

zakieh-tayyebi committed:
minor edit

sunwsusan commented on core/scripts/migrate_requests_to_new_base.py:
```diff
@@ -615,13 +620,20 @@ def create_requests_records(record: dict):
+        record.get("Request Types"),
+        record.get("Furniture Items"),
+        record.get("Kitchen Items"),
+        record.get("Bed Details"),
```
before concatenating, might want to check if the field exists and are not None.  the previous code used 

```
record.get("Request Types", [])
```

the pandas equivalent of that would be

```
record.get("Request Types") or pd.DataFrame()
```

etc.

zakieh-tayyebi replied:
`pd.concat` would have the same result if the list has an empty data frame, or if it has None (which is what `get` would return by default). Is there a difference?

sunwsusan replied:
a bit yeah..  an empty df work fine with pd.concat, but any None values will crash the `pd.concat()`.  if we are confident that we will always be able to `.get()` these values, then this is just nitpicking on my end and feel free to ignore.  otherwise, two ways to get around those could be:

option 1: handle None values prior to concatenating 
```
    request_types = record.get("Request Types") or pd.DataFrame()
    furniture_items = record.get("Furniture Items") or pd.DataFrame()
    kitchen_items = record.get("Kitchen Items") or pd.DataFrame()
    bed_details = record.get("Bed Details") or pd.DataFrame()

    all_reqs = pd.concat([
        request_types,
        furniture_items,
        kitchen_items,
        bed_details]
      , ignore_index=True)
```

or, option 2: filter out the None values and only keep not null dfs as an interim step 
```
    dataframes = [
        record.get("Request Types"),
        record.get("Furniture Items"),
        record.get("Kitchen Items"),
        record.get("Bed Details"),
    ]
    
    valid_dataframes = [df for df in dataframes if df is not None]
    
    if not valid_dataframes:
        all_reqs = pd.DataFrame()
    else:
        all_reqs = pd.concat(valid_dataframes, ignore_index=True)
```

wdyt?

zakieh-tayyebi replied:
That's a good point. Added default to `pd.DataFrame()` so similar to option 1.

sunwsusan commented on core/scripts/migrate_requests_to_new_base.py:
```diff
@@ -382,29 +392,30 @@ def transform_open_requests(
 
-    output = defaultdict(list)
-    for item in all_items:
+    output = defaultdict(pd.DataFrame)
```
weird pandas quirk.  

potential issue 1:
current code will create an empty pandas dataframe, but when you do your _first_ concatenation (e.g. `pd.concat([empty_df, new_df])`, pandas will return the new DataFrame (`new_df`) directly instead of concatenating, losing the first item

potential issue 2:
whenever the dataframe encounters a key that doesn't exist, `defaultdict` will automatically creates a new empty DataFrame.

to get around, this, please replace line 417 with the following, which will initialize a pandas dataframe with all expected output keys prior to loop:

```
output = {}
for sub_item in REQUEST_SUB_ITEMS:
    if sub_item["items_output_field"]:
        output[sub_item["items_output_field"]] = pd.DataFrame()
    if sub_item["request_type_output_field"]:
        output[sub_item["request_type_output_field"]] = pd.DataFrame()
output[new_field_name] = pd.DataFrame()
```

zakieh-tayyebi replied:
I might be misunderstanding, but what you described is what I intended so it would behave similarly to `defaultdict` in the original script:
regarding potential issue 1: Do we need to keep an empty data frame? In the original script, items are added to a vector that is initially empty. Here, rows are added to a data frame that is initially empty.
regarding potential issue 2: That is the intended result, similarly to the original script, we want to get an empty data frame to add to, and it's fine for that key to not exist otherwise.

sunwsusan replied:
oh oops I didn't realize that was the intended behavior.  then yeah this comment is unnecessary.  resolving!

sunwsusan commented on core/scripts/migrate_requests_to_new_base.py:
```diff
@@ -1,6 +1,9 @@
 from collections import defaultdict
 import copy
+from datetime import datetime
+from dateutil import parser as date_parser
```
nit: these are not currently being used, consider removing?

zakieh-tayyebi replied:
removed. thanks!

sunwsusan commented on core/scripts/migrate_requests_to_new_base.py:
```diff
@@ -382,29 +392,30 @@ def transform_open_requests(
+    ])
+
+    # pick oldest request of each type:
+    all_items_df = select_oldest_request(all_items_df)
```
currently, if `r.get(old_field_name)` returns None, the for loop will iterate over None and error out with TypeError.  

not the most concise workaround, open to better solutions:

```
    # remove  None/empty values before concatenation
    valid_items = []
    for r in records:
        field_value = r.get(old_field_name)
        if field_value and isinstance(field_value, list):
            for item in field_value:
                valid_items.append((r, item))
    
    if not valid_items:
        # if no valid data, return empty df
        return {
            new_field_name: pd.DataFrame(),
            "Bed Details": pd.DataFrame(),
            "Furniture Items": pd.DataFrame(),
            "Kitchen Items": pd.DataFrame(),
            "Social Service Requests": pd.DataFrame(),
        }

    all_items_df = pd.concat([
        item_df(r, item) for r, item in valid_items
    ])
```

sunwsusan replied:
fyi this might also be a nit, similar to [the thread above](https://github.com/bushwickayudamutua/bam-automation/pull/26#discussion_r2319377667). open to discussion

zakieh-tayyebi replied:
I updated this to return an empty list if the key doesn't exist, and create an empty data frame if the list is empty. But I think we can assume each household has at least one open request so this would not occur.

zakieh-tayyebi committed:
added pandas and removed unused libs

zakieh-tayyebi committed:
default to empty dataframe

sunwsusan reviewed (approved):
lgtm!

zakieh-tayyebi committed:
consider none list

zakieh-tayyebi committed:
Removed unnecessary deduping

zakieh-tayyebi committed:
fixed social services

zakieh-tayyebi committed:
fixed issue with date format

zakieh-tayyebi committed:
migrate address fields to requests table

jmasom committed:
Update/use orm in migration script (#33)

abelsonlive reviewed (approved):
LGTM!  The Airtable ORM stuff is really cool!

gkopplin commented on core/scripts/migrate_requests_to_new_base.py:
```diff
@@ -106,6 +103,15 @@ def select_first_non_null(
+        if record.get(old_field_name):
+            return {new_field_name: record.get(old_field_name)}
+    return {}
+
```
not blocking, but these helper functions would be good candidate for unit tests

gkopplin commented on core/scripts/migrate_requests_to_new_base.py:
```diff
@@ -614,85 +624,83 @@ def create_requests_records(record: dict):
-    return [r["id"] for r in requests_response]
+
+    request_records = request_records1 + request_records2
+    Request.batch_save(request_records)
```
what's the reason for the change from batch_create to batch_save? I wouldn't have expected the date changes to require this. Did we just not account for the possible need of running the script more than once when we chose batch_create or something?

jmasom replied:
i believe the relevant difference is that `batch_save()` takes models as opposed to dictionaries

gkopplin reviewed (commented):
Looks good, but I'll take another look after you resolve the merge conflicts.

Here's a few suggestions for things we could add to the description:
- context on the changes that aren't related to the migration script (or, even better, could we move those changes to a separate PR?)
- a note on the new ORM models
- adding that Bed Details will also have geocode data
- details of testing (since we're switching to dataframes, the diff is pretty large, which makes it harder to spot issues)

jmasom committed:
Merge branch 'main' into migration/add_date_to_requests

zakieh-tayyebi committed:
migrate Cita Availability into new separate fields

zakieh-tayyebi committed:
removing Form Submissions field

zakieh-tayyebi commented:
- The field "Cita Availability" has many different tags. We migrate "Needs Delivery" and "Needs Email Outreach" as boolean fields in the new base, and ignore other tags (for now).
- We do not need to migrate form submissions, so we deleted the field from the new base, and removed its migration from this script.

zakieh-tayyebi committed:
fixed cita availability bug

jmasom committed:
remove form_submission field from Household model

gkopplin reviewed (approved):
Looks good! Nice work!

zakieh-tayyebi committed:
migrate other languages field

zakieh-tayyebi committed:
convert to datetime date

zakieh-tayyebi committed:
added other languages to household class

zakieh-tayyebi commented:
Current status:
- Fixed issue with date formats
- Added "Other Languages" field to new base
- To fix: Email is not the most recent
